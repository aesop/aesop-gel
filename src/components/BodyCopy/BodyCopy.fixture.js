import { HYPERLINK_STYLE_TYPES } from '~/constants';

const { EXTERNAL_NO_ICON_TEXT_LINK } = HYPERLINK_STYLE_TYPES;

export default {
  eyebrow: 'Nullam id erat est',
  heading: 'Lorem ipsum dolor sit amet',
  subHeading: 'Nunc dignissim quis arcu a',
  theme: 'dark',
  cta: {
    id: '1',
    style: 'External Button Link',
    text: 'Nullam id erat est',
    children: 'Nullam id erat est',
    url: '/',
  },
  secondaryCta: {
    id: '2',
    style: 'External Button Link',
    text: 'Phasellus tempus',
    children: 'Phasellus tempus',
    url: '/',
  },
  links: [
    {
      id: 'Morbi-id-nulla',
      style: EXTERNAL_NO_ICON_TEXT_LINK,
      children: 'Morbi id nulla',
      url: 'https://aesop.com/',
    },
    {
      id: 'Nullam-ultrices',
      style: EXTERNAL_NO_ICON_TEXT_LINK,
      children: 'Nullam ultrices',
      url: 'https://aesop.com/',
    },
  ],

  copy:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam id erat est. Ut convallis, neque id semper fermentum, urna risus gravida nibh, vel molestie sapien leo ac tortor. Nunc dignissim quis arcu a lobortis. Phasellus tempus consequat erat, at egestas elit pulvinar lacinia. Nullam pretium, risus a dictum rhoncus, ex libero cursus erat, dignissim malesuada tortor justo quis enim.',
};
