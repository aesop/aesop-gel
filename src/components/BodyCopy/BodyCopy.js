import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import LinkButtonGroup from '~/components/LinkButtonGroup';
import Hyperlink from '~/components/Hyperlink';
import SectionHeading from '~/components/SectionHeading';
import SubNav from '~/components/SubNav';
import styles from './BodyCopy.module.css';

const BodyCopy = ({
  childrenClassNames,
  className,
  copy,
  cta,
  ctaTextAlign,
  eyebrow,
  hasSerifFontHeading,
  heading,
  id,
  links,
  renderCtaAsSubnav,
  secondaryCta,
  subHeading,
  theme,
}) => {
  const classSet = cx(styles.base, styles[theme], className);

  const LinkBlock = () => {
    return renderCtaAsSubnav ? (
      <SubNav
        links={[cta && { ...cta }, secondaryCta && { ...secondaryCta }].filter(
          Boolean
        )}
        theme={theme}
      />
    ) : (
      <LinkButtonGroup
        className={styles.ctaWrapper}
        textAlign={ctaTextAlign}
        theme={theme}
      >
        {cta && <Hyperlink {...cta}>{cta.text}</Hyperlink>}
        {secondaryCta && (
          <Hyperlink {...secondaryCta}>{secondaryCta.text}</Hyperlink>
        )}
      </LinkButtonGroup>
    );
  };

  return (
    <article className={classSet} id={id}>
      <SectionHeading
        childrenClassNames={{
          eyebrow: cx(styles.eyebrow, childrenClassNames.eyebrow),
          heading: cx(styles.heading, childrenClassNames.heading),
          subHeading: cx(styles.subHeading, childrenClassNames.subHeading),
        }}
        eyebrow={eyebrow}
        hasSerifFontHeading={hasSerifFontHeading}
        heading={heading}
        id={id}
        subHeading={subHeading}
        theme={theme}
      />

      {copy && <div className={styles.copy}>{copy}</div>}

      {links ? <SubNav links={links} theme={theme} /> : <LinkBlock />}
    </article>
  );
};

BodyCopy.propTypes = {
  childrenClassNames: PropTypes.shape({
    eyebrow: PropTypes.string,
    heading: PropTypes.string,
    subHeading: PropTypes.string,
  }),
  className: PropTypes.string,
  copy: PropTypes.node,
  cta: PropTypes.object,
  ctaTextAlign: PropTypes.string,
  eyebrow: PropTypes.string,
  hasSerifFontHeading: PropTypes.bool,
  heading: PropTypes.string,
  id: PropTypes.string,
  links: PropTypes.arrayOf(PropTypes.object),
  renderCtaAsSubnav: PropTypes.bool,
  secondaryCta: PropTypes.object,
  subHeading: PropTypes.string,
  theme: PropTypes.oneOf(['dark', 'light']),
};

BodyCopy.defaultProps = {
  childrenClassNames: {
    eyebrow: undefined,
    heading: undefined,
    subHeading: undefined,
  },
  className: undefined,
  copy: undefined,
  cta: undefined,
  ctaTextAlign: undefined,
  eyebrow: undefined,
  hasSerifFontHeading: false,
  heading: undefined,
  id: undefined,
  links: undefined,
  renderCtaAsSubnav: false,
  secondaryCta: undefined,
  subHeading: undefined,
  theme: 'dark',
};

export default BodyCopy;
