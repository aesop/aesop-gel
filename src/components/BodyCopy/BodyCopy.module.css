@import 'styles/variables/index';

.base {
  padding: 20px;
  color: var(--color-body-copy);
  font-family: var(--font-family-sans-serif-regular);
}

.hasLightText {
  color: var(--color-white);
}

.TextOverHero {
  .heading,
  .subHeading {
    margin: 0;
  }

  .subHeading {
    padding-top: 5px;
  }

  .copy {
    margin: 30px 0 0;
  }

  a {
    width: 100%;
    margin-top: 30px;

    @media (--viewport-md) {
      width: auto;
    }
  }
}

.TwoColumn {
  .heading {
    font-family: var(--font-family-zapf-serif);
  }
}

.HalfWidthFullBleedHero {
  .eyebrow {
    padding: 0 0 19px;

    @media (--viewport-lg) {
      padding: 0 0 14px;
      margin-top: 60px;
    }

    @media (--viewport-xl) {
      padding: 0 0 19px;
    }
  }

  .heading {
    padding: 0 0 17px 0;
    border-bottom: 2px solid var(--color-grey-10);
    font-family: var(--font-family-zapf-serif);

    @media (--viewport-md) {
      padding: 0 0 12px;
    }

    @media (--viewport-xl) {
      padding: 0 0 17px;
    }
  }
}

.copy p {
  margin: 10px 0;
  font-size: rem(17px);
  line-height: 1.7;

  &:first-of-type {
    margin-top: 0;
  }

  &:last-of-type {
    margin-bottom: 0;
  }

  a {
    color: var(--color-grey-20);
    font-weight: var(--font-weight-normal);
    text-decoration: underline;
  }
}

.quote {
  margin-bottom: 30px;
  font-family: var(--font-family-zapf-serif);
  font-size: rem(26px);
  line-height: 1.5;
}

.ctaWrapper {
  margin-top: 20px;
}
