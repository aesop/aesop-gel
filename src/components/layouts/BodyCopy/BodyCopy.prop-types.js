import PropTypes from 'prop-types';
import { getPropTypeOptionValues } from '~/utils/propTypes';

export const HYPERLINK_PROP_TYPE_SHAPE = PropTypes.object;
export const PARENT_PROP_TYPE_OPTIONS = {
  TEXT_OVER_HERO: 'TextOverHero',
  HALF_WIDTH_FULL_BLEED: 'HalfWidthFullBleed',
};

export const PROP_TYPES = {
  className: PropTypes.string,
  copy: PropTypes.oneOf([PropTypes.string, PropTypes.node]),
  cta: HYPERLINK_PROP_TYPE_SHAPE, // @TODO replace with HyperLinkType
  eyebrow: PropTypes.string,
  id: PropTypes.string,
  parent: PropTypes.oneOf(getPropTypeOptionValues(PARENT_PROP_TYPE_OPTIONS)),
  primaryTitle: PropTypes.string,
  quote: PropTypes.string,
  secondaryTitle: PropTypes.string,
};

export default PROP_TYPES;
