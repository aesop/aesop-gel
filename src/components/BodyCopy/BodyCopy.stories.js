import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';
import '~.storybook/storybook.module.css';
import BodyCopy from './BodyCopy';

storiesOf('BodyCopy', module).add('Base component', () => (
  <BodyCopy
    copy={text(
      'copy',
      'Parsley Seed Cream. [The Paris Review](http://theparisreview.org) Intensely soothing, nourishing and hydrating, this elegant formulation contains a potent blend of botanicals that offer fortification of the highest order against free radicals.',
    )}
    cta={{
      id: text('cta.id', 'button-id'),
      text: text('cta.text', 'Read more'),
      url: text('cta.url', 'http://aesop.com'),
      openInANewWindow: boolean('cta.openInANewWindow', true),
    }}
    eyebrow={text('eyebrow', 'Recommended Nearby')}
    id={text('id', 'Anti-Oxidant')}
    primaryTitle={text(
      'primaryTitle',
      'Aesop and The Paris Review: A Partnership Extended',
    )} // TBC -- nested component
    secondaryTitle={text('secondaryTitle', 'Active Nutrients')}
  />
));
