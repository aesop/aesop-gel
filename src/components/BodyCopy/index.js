/* istanbul ignore file */
import { BodyCopy } from './BodyCopy';

export { BodyCopy };
