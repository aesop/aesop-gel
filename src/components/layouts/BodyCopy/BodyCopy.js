import React from 'react';
import cx from 'classnames';
import marked from 'marked';
import parse from 'html-react-parser';
import Heading from '~/components/elements/Heading/Heading';
import Hyperlink from '~/components/elements/Hyperlink';
import PROP_TYPES from './BodyCopy.prop-types';
import { getPrimaryTitleLevel, getSecondaryTitleLevel } from './BodyCopy.utils';
import styles from './BodyCopy.module.css';

const BodyCopy = ({
  className,
  copy,
  cta,
  eyebrow,
  id,
  parent = '',
  primaryTitle,
  quote,
  secondaryTitle,
}) => {
  const classSet = cx(styles.base, className, styles[parent]);

  return (
    <section className={classSet} id={id}>
      {eyebrow && (
        <Heading className={styles.eyebrow} level="2" size="xSmall">
          {eyebrow}
        </Heading>
      )}

      {primaryTitle && (
        <Heading
          className={styles.primaryTitle}
          level={getPrimaryTitleLevel(eyebrow)}
          size="xLarge"
        >
          {primaryTitle}
        </Heading>
      )}

      {secondaryTitle && (
        <Heading
          className={styles.secondaryTitle}
          level={getSecondaryTitleLevel(eyebrow, primaryTitle)}
          size="small"
        >
          {secondaryTitle}
        </Heading>
      )}

      {copy && (
        <div className={styles.copy}>{parse(marked(copy.toString()))}</div>
      )}

      {cta && cta.text && (
        <Hyperlink
          openInANewWindow={cta.openInANewWindow}
          style={cta.style}
          url={cta.url}
        >
          {cta.text}
        </Hyperlink>
      )}

      {quote && <p className={styles.quote}>{quote}</p>}
    </section>
  );
};

BodyCopy.propTypes = PROP_TYPES;

export default BodyCopy;
