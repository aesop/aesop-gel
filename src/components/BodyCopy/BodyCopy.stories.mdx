import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { ThemeContextProvider } from '~/contexts';
import { LinkButtonGroup } from '~/components/LinkButtonGroup';
import { SubNav } from '~/components/SubNav';
import { BodyCopy } from './BodyCopy';
import { BodyCopyFixture } from './BodyCopy.fixture';

<Meta title="Components/BodyCopy" component={BodyCopy} />

# BodyCopy

<Canvas>
  <Story name="Base component">
    {(() => {
      const theme = knobs.select(
        'theme',
        ['dark', 'light'],
        BodyCopyFixture.theme,
        'Presentation',
      );
      return (
        <ThemeContextProvider theme={theme}>
          <BodyCopy
            copy={knobs.text('copy', BodyCopyFixture.copy, 'Copy')}
            eyebrow={knobs.text('eyebrow', BodyCopyFixture.eyebrow, 'Copy')}
            hasFlushHeading={knobs.boolean(
              'hasFlushHeading',
              false,
              'Presentation',
            )}
            hasSerifFontHeading={knobs.boolean(
              'hasSerifFontHeading',
              false,
              'Typography',
            )}
            heading={knobs.text('heading', BodyCopyFixture.heading, 'Copy')}
            subHeading={knobs.text(
              'subHeading',
              BodyCopyFixture.subHeading,
              'Copy',
            )}
            content={(() => {
              const options = {
                SubNav: {
                  Component: SubNav,
                  props: BodyCopyFixture.links,
                },
                LinkButtonGroup: {
                  Component: LinkButtonGroup,
                  props: BodyCopyFixture.linkButtonGroup,
                },
              };
              const componentName = knobs.select(
                'content',
                Object.keys(options),
                'LinkButtonGroup',
              );
              const { Component, props } = options[componentName];
              return <Component {...props} />;
            })()}
          />
        </ThemeContextProvider>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={BodyCopy} />
