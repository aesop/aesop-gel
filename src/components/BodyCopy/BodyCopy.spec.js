import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import BodyCopy from './BodyCopy';
import BodyCopyFixture from './BodyCopy.fixture';

configure({ adapter: new Adapter() });

describe('<BodyCopy />', () => {
  it('should be defined', () => {
    expect(BodyCopy).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <BodyCopy
          copy="Parsley Seed Cream. [The Paris Review](http://theparisreview.org) Intensely soothing, nourishing and hydrating, this elegant formulation contains a potent blend of botanicals that offer fortification of the highest order against free radicals."
          cta={BodyCopyFixture.cta}
          eyebrow="Recommended Nearby"
          heading="Aesop & The Paris Review: A Partnership Extended"
          id="Anti-Oxidant"
          subHeading="Active Nutrients"
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders promary title as an `h2` if no `eyebrow` prop is passed', () => {
    const tree = renderer
      .create(
        <BodyCopy heading="Aesop & The Paris Review: A Partnership Extended" />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders promary title as an `h2` if no `eyebrow` or `primaryTitle` prop are passed', () => {
    const tree = renderer
      .create(
        <BodyCopy subHeading="Aesop & The Paris Review: A Partnership Extended" />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
