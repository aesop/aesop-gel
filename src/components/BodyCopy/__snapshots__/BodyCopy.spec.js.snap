// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BodyCopy /> renders base component correctly 1`] = `
<article
  className="base dark undefined"
  id="Anti-Oxidant"
>
  <header
    className="base"
    id="Anti-Oxidant"
  >
    <h2
      className="base mediumWeightFont xXSmall dark eyebrow"
    >
      Recommended Nearby
    </h2>
    <h3
      className="base xLarge dark heading"
    >
      Aesop & The Paris Review: A Partnership Extended
    </h3>
    <h4
      className="base mediumWeightFont xSmall dark subHeading"
    >
      Active Nutrients
    </h4>
  </header>
  <div
    className="copy dark"
  >
    Parsley Seed Cream. [The Paris Review](http://theparisreview.org) Intensely soothing, nourishing and hydrating, this elegant formulation contains a potent blend of botanicals that offer fortification of the highest order against free radicals.
  </div>
  <div
    className="base flush"
  />
</article>
`;

exports[`<BodyCopy /> renders promary title as an \`h2\` if no \`eyebrow\` or \`primaryTitle\` prop are passed 1`] = `
<article
  className="base dark undefined"
>
  <header
    className="base"
  >
    <h2
      className="base mediumWeightFont xSmall dark subHeading"
    >
      Aesop & The Paris Review: A Partnership Extended
    </h2>
  </header>
  <div
    className="base flush"
  />
</article>
`;

exports[`<BodyCopy /> renders promary title as an \`h2\` if no \`eyebrow\` prop is passed 1`] = `
<article
  className="base dark undefined"
>
  <header
    className="base"
  >
    <h2
      className="base xLarge dark heading"
    >
      Aesop & The Paris Review: A Partnership Extended
    </h2>
  </header>
  <div
    className="base flush"
  />
</article>
`;
